src/test/java/io/github/mapepire_ibmi/PoolTest.java

```diff
@@ -11,49 +15,273 @@
+        assertTrue(endNoPool - startNoPool > endPool2 - startPool2);
+    }
+
+    @Test
+    void popJobGivesReadyJob() throws Exception {
+        PoolOptions options = new PoolOptions(MapepireTest.getCreds(), 5, 5);
+        Pool pool = new Pool(options);
+        pool.init().get();
+        assertEquals(5, pool.getActiveJobCount());
+
+        List<CompletableFuture<QueryResult<Object>>> futures = new ArrayList<>();
+        for (int i = 0; i < 2; i++) {
+            futures.add(pool.execute("SELECT * FROM SAMPLE.SYSCOLUMNS"));
+        }
+        CompletableFuture<Void> allFutures = CompletableFuture.allOf(futures.toArray(new CompletableFuture[0]));
+
+        SqlJob job = pool.popJob().get();
+        assertEquals(JobStatus.Ready, job.getStatus());
+        assertEquals(0, job.getRunningCount());
+        assertEquals(4, pool.getActiveJobCount());
+
+        allFutures.join();
+        pool.end();
+    }
+
+    @Test
+    void popJobGivesNewJob() throws Exception {
+        PoolOptions options = new PoolOptions(MapepireTest.getCreds(), 1, 1);
+        Pool pool = new Pool(options);
+        pool.init().get();
+        assertEquals(1, pool.getActiveJobCount());
+
+        CompletableFuture<QueryResult<Object>> future = pool.execute("SELECT * FROM SAMPLE.SYSCOLUMNS");
+
+        SqlJob job = pool.popJob().get();
+        assertEquals(JobStatus.Ready, job.getStatus());
+        assertEquals(0, job.getRunningCount());
+        assertEquals(1, pool.getActiveJobCount());
+
+        future.join();
+        pool.end();
+    }
+
+    @Test
+    void getJobGivesReadyJob() throws Exception {
+        PoolOptions options = new PoolOptions(MapepireTest.getCreds(), 2, 2);
+        Pool poolSpy = spy(new Pool(options));
+        poolSpy.init().get();
+        assertEquals(2, poolSpy.getActiveJobCount());
+
+        CompletableFuture<QueryResult<Object>> future = poolSpy.execute("SELECT * FROM SAMPLE.SYSCOLUMNS");
+
+        SqlJob job = poolSpy.getJob();
+        assertEquals(JobStatus.Ready, job.getStatus());
+        assertEquals(0, job.getRunningCount());
+        assertEquals(2, poolSpy.getActiveJobCount());
+        verify(poolSpy, times(0)).addJob();
+
+        future.join();
+        poolSpy.end();
+    }
+
+    @Test
+    void getJobGivesFreeJob() throws Exception {
+        PoolOptions options = new PoolOptions(MapepireTest.getCreds(), 2, 2);
+        Pool poolSpy = spy(new Pool(options));
+        poolSpy.init().get();
+        assertEquals(2, poolSpy.getActiveJobCount());
+
+        List<CompletableFuture<QueryResult<Object>>> futures = new ArrayList<>();
+        for (int i = 0; i < 3; i++) {
+            futures.add(poolSpy.execute("SELECT * FROM SAMPLE.SYSCOLUMNS"));
+        }
+        CompletableFuture<Void> allFutures = CompletableFuture.allOf(futures.toArray(new CompletableFuture[0]));
+
+        SqlJob job = poolSpy.popJob().get();
+        assertEquals(JobStatus.Ready, job.getStatus());
+        assertEquals(1, job.getRunningCount());
```

Comment:
I'm wondering how the running count can be 1, but at the same time the job is ready?

Reply:
I will make these fixes in a separate PR once https://github.com/Mapepire-IBMi/mapepire-java/issues/43 is fixed